api/pkg/inference-logger/logger/mlobs_sink.go

```diff
@@ -0,0 +1,154 @@
+
+	return sink
+}
+
+func (m *MLObsSink) newPredictionLog(rawLogEntry *LogEntry) (*upiv1.PredictionLog, error) {
+	predictionLog := &upiv1.PredictionLog{}
+	standardModelRequest := &StandardModelRequest{}
+	err := json.Unmarshal(rawLogEntry.RequestPayload.Body, standardModelRequest)
+	if err != nil {
+		return nil, err
+	}
+	standardModelResponse := &StandardModelResponse{}
+	err = json.Unmarshal(rawLogEntry.ResponsePayload.Body, standardModelResponse)
+	if err != nil {
+		return nil, err
+	}
+	predictionLog.PredictionId = standardModelRequest.SessionId
+	predictionLog.ModelName = m.modelName
+	predictionLog.ModelVersion = m.modelVersion
+	predictionLog.ProjectName = m.projectName
+	featuresTable := &structpb.Struct{
+		Fields: make(map[string]*structpb.Value),
+	}
+	predictionTable := &structpb.Struct{
+		Fields: make(map[string]*structpb.Value),
+	}
+	featureTableData := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelRequest.Instances)),
+	}
+	predictionTableData := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelResponse.Predictions)),
+	}
+	rowIds := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelRequest.RowIds)),
+	}
+	for i, instance := range standardModelRequest.Instances {
+		featureTableRow := &structpb.ListValue{
+			Values: make([]*structpb.Value, len(instance)),
+		}
+		for j, value := range instance {
+			if value != nil {
+				featureTableRow.Values[j] = structpb.NewNumberValue(*value)
+			} else {
+				featureTableRow.Values[j] = structpb.NewNullValue()
+			}
+		}
+
+		featureTableData.Values[i] = structpb.NewListValue(featureTableRow)
+		predictionTableData.Values[i] = structpb.NewNumberValue(standardModelResponse.Predictions[i])
+		rowIds.Values[i] = structpb.NewStringValue(standardModelRequest.RowIds[i])
+	}
+	featuresTable.Fields["data"] = structpb.NewListValue(featureTableData)
+	featuresTable.Fields["row_ids"] = structpb.NewListValue(rowIds)
+	predictionTable.Fields["data"] = structpb.NewListValue(predictionTableData)
+	predictionTable.Fields["row_ids"] = structpb.NewListValue(rowIds)
+
+	predictionLog.Input = &upiv1.ModelInput{
+		FeaturesTable: featuresTable,
+	}
+	predictionLog.Output = &upiv1.ModelOutput{
+		PredictionResultsTable: predictionTable,
+	}
+	return predictionLog, nil
+}
+
+func (m *MLObsSink) buildNewKafkaMessage(predictionLog *upiv1.PredictionLog) (*kafka.Message, error) {
+	logBytes, err := proto.Marshal(predictionLog)
+	if err != nil {
+		return nil, err
+	}
+	topic := fmt.Sprintf("caraml-%s-%s-%s-prediction-log", m.projectName, m.modelName, m.modelVersion)
+	return &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic: &topic,
+		},
+		Value: logBytes,
+	}, nil
+}
+
+func (m *MLObsSink) Sink(rawLogEntries []*LogEntry) error {
+	for _, rawLogEntry := range rawLogEntries {
+		predictionLog, err := m.newPredictionLog(rawLogEntry)
+		if err != nil {
+			m.logger.Errorf("unable to convert log entry: %v", err)
```

Comment:
Should we return err here and cancel the sinking for invalid entry? Same as below error scenarios.

Reply:
I thought about this, but that will mean that the model itself will goes down if there are any issue with logging, which i am not sure if it is a desirable behaviour. The current inference log logger also print out the error log as opposed to returning error.